reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -56,87 +64,73 @@ public int getPrefetch() {
         @SuppressWarnings("rawtypes")
-        static final AtomicReferenceFieldUpdater<AbstractSwitchOnFirstInner, CoreSubscriber> INNER =
-                AtomicReferenceFieldUpdater.newUpdater(AbstractSwitchOnFirstInner.class, CoreSubscriber.class, "inner");
+        static final AtomicReferenceFieldUpdater<AbstractSwitchOnFirstMain, CoreSubscriber> INNER =
+                AtomicReferenceFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, CoreSubscriber.class, "inner");
 
         volatile int wip;
         @SuppressWarnings("rawtypes")
-        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstInner> WIP =
-                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstInner.class, "wip");
+        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstMain> WIP =
+                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, "wip");
 
         volatile int state;
         @SuppressWarnings("rawtypes")
-        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstInner> STATE =
-                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstInner.class, "state");
+        static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstMain> STATE =
+                AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, "state");
 
         @SuppressWarnings("unchecked")
-        AbstractSwitchOnFirstInner(
+        AbstractSwitchOnFirstMain(
                 CoreSubscriber<? super R> outer,
-                BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer) {
+                BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer,
+                boolean cancelSourceOnComplete) {
             this.outer = outer instanceof Fuseable.ConditionalSubscriber
-                ? new SwitchOnFirstConditionalInnerSubscriber<>(this, (Fuseable.ConditionalSubscriber<R>) outer)
-                : new SwitchOnFirstInnerSubscriber<>(this, outer);
+                ? new SwitchOnFirstConditionalControlSubscriber<>(this, (Fuseable.ConditionalSubscriber<R>) outer, cancelSourceOnComplete)
+                : new SwitchOnFirstControlSubscriber<>(this, outer, cancelSourceOnComplete);
             this.transformer = transformer;
         }
 
         @Override
         @Nullable
         public Object scanUnsafe(Attr key) {
-            if (key == Attr.CANCELLED) return cancelled;
-            if (key == Attr.TERMINATED) return done || cancelled;
+            final boolean isCancelled = this.inner == Operators.emptySubscriber();
+
+            if (key == Attr.CANCELLED) return isCancelled && !this.done;
+            if (key == Attr.TERMINATED) return this.done || isCancelled;
 
             return InnerOperator.super.scanUnsafe(key);
         }
 
         @Override
         public CoreSubscriber<? super R> actual() {
-            return outer;
-        }
-
-        @Override
-        public Context currentContext() {
```

Comment:
@simonbasle I ended up with the conclusion that context can not be mutable for upstream, thus removed that tricky part

Reply:
this boils down to `outer.currentContext()`, so ok